src/backend/distributed/master/master_delete_protocol.c

```diff
@@ -359,99 +365,169 @@ static int
-								 quotedShardName);
+				/*
+				 * The active DROP SCHEMA/DATABASE ... CASCADE will drop the
+				 * shard, if we try to drop it over another connection, we will
+				 * get into a distributed deadlock.
+				 */
 			}
-			else if (storageType == SHARD_STORAGE_COLUMNAR ||
-					 storageType == SHARD_STORAGE_FOREIGN)
+			else
 			{
-				appendStringInfo(workerDropQuery, DROP_FOREIGN_TABLE_COMMAND,
-								 quotedShardName);
+				char storageType = shardInterval->storageType;
+
+				/* build command to drop the shard */
+				const char *dropShardPlacementCommand = CreateDropShardPlacementCommand(
+					schemaName, shardRelationName, storageType);
+
+				/*
+				 * Try to open a new connection (or use an existing one) to
+				 * connect to target node to drop shard placement over that
+				 * remote connection
+				 */
+				ExecuteDropShardPlacementCommandRemotely(shardPlacement,
+														 shardRelationName,
+														 dropShardPlacementCommand);
 			}
 
 			/*
-			 * The active DROP SCHEMA/DATABASE ... CASCADE will drop the shard, if we
-			 * try to drop it over another connection, we will get into a distributed
-			 * deadlock.
+			 * Regardless of that we dropped the placement or not, delete the
+			 * placement metadadata.
 			 */
-			if (shardPlacement->groupId == COORDINATOR_GROUP_ID &&
-				IsCoordinator() &&
-				DropSchemaOrDBInProgress())
-			{
-				DeleteShardPlacementRow(shardPlacement->placementId);
-				continue;
-			}
+			DeleteShardPlacementRow(shardPlacementId);
+		}
 
-			MultiConnection *connection = GetPlacementConnection(connectionFlags,
-																 shardPlacement,
-																 NULL);
+		/*
+		 * Now that we deleted all placements of the shard (or their metadata),
+		 * delete the shard metadata as well.
+		 */
+		DeleteShardRow(shardId);
+	}
 
-			RemoteTransactionBeginIfNecessary(connection);
+	int droppedShardCount = list_length(deletableShardIntervalList);
 
-			if (PQstatus(connection->pgConn) != CONNECTION_OK)
-			{
-				uint64 placementId = shardPlacement->placementId;
+	return droppedShardCount;
+}
 
-				ereport(WARNING, (errmsg("could not connect to shard \"%s\" on node "
-										 "\"%s:%u\"", shardRelationName, workerName,
-										 workerPort),
-								  errdetail("Marking this shard placement for "
-											"deletion")));
 
-				UpdateShardPlacementState(placementId, SHARD_STATE_TO_DELETE);
+/*
+ * ExecuteDropShardPlacementCommandRemotely executes the given DROP shard command
+ * via remote critical connection.
+ */
+static void
+ExecuteDropShardPlacementCommandRemotely(ShardPlacement *shardPlacement, const
+										 char *shardRelationName, const
+										 char *dropShardPlacementCommand)
+{
+	Assert(shardPlacement != NULL);
+	Assert(shardRelationName != NULL);
+	Assert(dropShardPlacementCommand != NULL);
 
-				continue;
-			}
+	uint32 dropShardConnectionFlags = FOR_DDL;
+	MultiConnection *connection = GetPlacementConnection(dropShardConnectionFlags,
```

Comment:
we typically always use `connectionFlags`, so `dropShard ` prefix seems a little unnecessary

Reply:
makes sense, reverting